File: class20/python/PES2001-Hexagonal.py
"""
PES2001 - Arquitetura Hexagonal (Ports & Adapters)
Aula 20: Clean Architecture, Hexagonal e Onion
Projeto e Engenharia de Software - Prof. Giovane Barcelos

Implementa Ports & Adapters com o domínio "Locadora de Veículos":
uma porta (VeiculoRepository), dois adaptadores (MySQL simulado e
memória) e um serviço de domínio que depende apenas da porta,
recebendo o adaptador concreto via injeção de dependência (DIP).
"""

from abc import ABC, abstractmethod


# --- Domínio: entidade ---
class Veiculo:
    def __init__(self, placa, modelo, disponivel=True):
        self.placa = placa
        self.modelo = modelo
        self.disponivel = disponivel

    def __repr__(self):
        status = "disponível" if self.disponivel else "locado"
        return f"{self.placa} - {self.modelo} ({status})"


# --- PORTA (interface no domínio) — secundária/driven ---
class VeiculoRepository(ABC):
    @abstractmethod
    def salvar(self, veiculo) -> None:
        pass

    @abstractmethod
    def buscar_por_placa(self, placa: str):
        pass

    @abstractmethod
    def listar_disponiveis(self) -> list:
        pass


# --- ADAPTADOR — implementação concreta (infraestrutura) ---
class VeiculoMySQLRepository(VeiculoRepository):
    def salvar(self, veiculo):
        print(f"[MySQL] INSERT INTO veiculos ... {veiculo.placa}")

    def buscar_por_placa(self, placa):
        print(f"[MySQL] SELECT ... WHERE placa='{placa}'")
        return None

    def listar_disponiveis(self):
        print("[MySQL] SELECT ... WHERE disponivel=true")
        return []


# --- ADAPTADOR — em memória, usado para testes (sem infraestrutura real) ---
class VeiculoMemoriaRepository(VeiculoRepository):
    def __init__(self):
        self._dados = {}

    def salvar(self, veiculo):
        self._dados[veiculo.placa] = veiculo

    def buscar_por_placa(self, placa):
        return self._dados.get(placa)

    def listar_disponiveis(self):
        return [v for v in self._dados.values() if v.disponivel]


# --- DOMÍNIO — não sabe nada sobre MySQL ou memória! ---
class LocadoraService:
    def __init__(self, repo: VeiculoRepository):
        self.repo = repo  # DIP: depende da porta, não do adaptador

    def cadastrar(self, veiculo):
        self.repo.salvar(veiculo)

    def veiculos_disponiveis(self):
        return self.repo.listar_disponiveis()


if __name__ == "__main__":
    print("=" * 60)
    print("  HEXAGONAL (PORTS & ADAPTERS) — LOCADORA DE VEÍCULOS")
    print("=" * 60)

    print("\n  --- Adaptador MySQL (simulado) ---")
    service_mysql = LocadoraService(VeiculoMySQLRepository())
    service_mysql.cadastrar(Veiculo("ABC1234", "Fusca"))

    print("\n  --- Adaptador em Memória (testes, sem infraestrutura real) ---")
    service_memoria = LocadoraService(VeiculoMemoriaRepository())
    service_memoria.cadastrar(Veiculo("XYZ9876", "Gol"))
    service_memoria.cadastrar(Veiculo("DEF4567", "Onix", disponivel=False))

    disponiveis = service_memoria.veiculos_disponiveis()
    print(f"  Veículos disponíveis: {len(disponiveis)}")
    for v in disponiveis:
        print(f"    {v}")

    assert len(disponiveis) == 1

    print("\n✓ Porta (VeiculoRepository): contrato definido pelo domínio.")
    print("✓ Adaptadores: MySQL e Memória são intercambiáveis via DIP.")
    print("✓ Trocar o adaptador não muda uma linha do domínio (LocadoraService).")

File: class20/java/PES2001-Hexagonal.java
/*
 * PES2001 - Arquitetura Hexagonal (Ports & Adapters)
 * Aula 20: Clean Architecture, Hexagonal e Onion
 * Projeto e Engenharia de Software - Prof. Giovane Barcelos
 *
 * Implementa Ports & Adapters com o domínio "Locadora de Veículos":
 * uma porta (VeiculoRepository), dois adaptadores (MySQL simulado e
 * memória) e um serviço de domínio que depende apenas da porta,
 * recebendo o adaptador concreto via injeção de dependência (DIP).
 *
 * Compilar/executar:
 *   javac PES2001-Hexagonal.java && java Hexagonal
 */

import java.util.*;

// --- Domínio: entidade ---
class Veiculo {
    String placa;
    String modelo;
    boolean disponivel;

    Veiculo(String placa, String modelo) {
        this(placa, modelo, true);
    }

    Veiculo(String placa, String modelo, boolean disponivel) {
        this.placa = placa;
        this.modelo = modelo;
        this.disponivel = disponivel;
    }

    @Override
    public String toString() {
        return String.format("%s - %s (%s)", placa, modelo, disponivel ? "disponível" : "locado");
    }
}

// --- PORTA (interface no domínio) ---
interface VeiculoRepository {
    void salvar(Veiculo veiculo);
    Veiculo buscarPorPlaca(String placa);
    List<Veiculo> listarDisponiveis();
}

// --- ADAPTADOR — implementação concreta (infraestrutura) ---
class VeiculoMySQLRepository implements VeiculoRepository {
    @Override
    public void salvar(Veiculo veiculo) {
        System.out.println("[MySQL] INSERT INTO veiculos ... " + veiculo.placa);
    }

    @Override
    public Veiculo buscarPorPlaca(String placa) {
        System.out.println("[MySQL] SELECT ... WHERE placa='" + placa + "'");
        return null;
    }

    @Override
    public List<Veiculo> listarDisponiveis() {
        System.out.println("[MySQL] SELECT ... WHERE disponivel=true");
        return new ArrayList<>();
    }
}

// --- ADAPTADOR — em memória, usado para testes (sem infraestrutura real) ---
class VeiculoMemoriaRepository implements VeiculoRepository {
    private final Map<String, Veiculo> dados = new HashMap<>();

    @Override
    public void salvar(Veiculo veiculo) {
        dados.put(veiculo.placa, veiculo);
    }

    @Override
    public Veiculo buscarPorPlaca(String placa) {
        return dados.get(placa);
    }

    @Override
    public List<Veiculo> listarDisponiveis() {
        List<Veiculo> resultado = new ArrayList<>();
        for (Veiculo v : dados.values()) {
            if (v.disponivel) resultado.add(v);
        }
        return resultado;
    }
}

// --- DOMÍNIO — não sabe nada sobre MySQL ou memória! ---
class LocadoraService {
    private final VeiculoRepository repo;

    LocadoraService(VeiculoRepository repo) {
        this.repo = repo; // DIP: depende da porta, não do adaptador
    }

    void cadastrar(Veiculo veiculo) {
        repo.salvar(veiculo);
    }

    List<Veiculo> veiculosDisponiveis() {
        return repo.listarDisponiveis();
    }
}

class Hexagonal {
    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  HEXAGONAL (PORTS & ADAPTERS) — LOCADORA DE VEÍCULOS");
        System.out.println("=".repeat(60));

        System.out.println("\n  --- Adaptador MySQL (simulado) ---");
        LocadoraService serviceMysql = new LocadoraService(new VeiculoMySQLRepository());
        serviceMysql.cadastrar(new Veiculo("ABC1234", "Fusca"));

        System.out.println("\n  --- Adaptador em Memória (testes, sem infraestrutura real) ---");
        LocadoraService serviceMemoria = new LocadoraService(new VeiculoMemoriaRepository());
        serviceMemoria.cadastrar(new Veiculo("XYZ9876", "Gol"));
        serviceMemoria.cadastrar(new Veiculo("DEF4567", "Onix", false));

        List<Veiculo> disponiveis = serviceMemoria.veiculosDisponiveis();
        System.out.println("  Veículos disponíveis: " + disponiveis.size());
        for (Veiculo v : disponiveis) System.out.println("    " + v);

        assert disponiveis.size() == 1;

        System.out.println("\n✓ Porta (VeiculoRepository): contrato definido pelo domínio.");
        System.out.println("✓ Adaptadores: MySQL e Memória são intercambiáveis via DIP.");
        System.out.println("✓ Trocar o adaptador não muda uma linha do domínio (LocadoraService).");
    }
}
